File: test/oa_config_test.R
# oa_config_test.R
#
# oa.config package test.
#
# Author: Derek Huang
# Copyright: MIT License
#

library(oa.config)
library(oa.testutils)

# shared VersionInfo object for the config library
cvi <- VersionInfo(library_path())

# test that library_dir() and data_dir() are the same
test_library_data_dirs <- function() {
  oa.tu_expect_eq(library_dir(), data_dir())
}

# test that VersionInfo$company() is the expected "OA Developers" string
test_cvi_company <- function() {
  oa.tu_expect_eq("OA Developers", cvi$company())
}

# test that VersionInfo$filename() is the file name from library_path()
test_cvi_filename <- function() {
  # path to active oa_config library
  path <- library_path()
  # oa_config filename from library_path() substring + from VersionInfo
  oa.tu_expect_eq(
    substring(path, regexpr("[/\\][^/\\]+$", path) + 1),
    cvi$filename()
  )
}

# test that VersionInfo$product() is the expected "OA Software Suite" string
test_cvi_product <- function() {
  oa.tu_expect_eq("OA Software Suite", cvi$product())
}

# test that VersionInfo$version_num() components are as expected
test_cvi_version_num <- function() {
  # SWIG wrapper produced an S4 class type since R only has 32-bit signed int
  vn <- cvi$version_num()
  # check each component is nonnegative
  oa.tu_expect_ge(vn$w1(), 0) &&
  oa.tu_expect_ge(vn$w2(), 0) &&
  oa.tu_expect_ge(vn$w3(), 0) &&
  oa.tu_expect_ge(vn$w4(), 0)
}

# named vector of tests to run
tests = c(
  test_library_data_dirs = test_library_data_dirs,
  test_cvi_filename = test_cvi_filename,
  test_cvi_company = test_cvi_company,
  test_cvi_product = test_cvi_product,
  test_cvi_version_num = test_cvi_version_num
)

# main
oa.tu_exit(oa.tu_main(desc = "oa_config unit tests.", tests = tests))

File: test/oa_config_test.py
"""oa_config module test.

This uses unittest to test wrapped oa_config classes and functions. To list the
test cases in this file, the --list-tests option can be used; this is mostly
useful for test harness integration, e.g. so CTest can discover the tests.

.. codeauthor:: Derek Huang <djh458@stern.nyu.edu>
"""

from pathlib import Path
import unittest

from oa_config import (
    VersionInfo,
    __file__ as oa_config_path,
    data_dir,
    library_dir,
    library_path
)
from oa_testutils import test_main

# location of the oa_config module
_oa_config_dir = Path(oa_config_path).parent


class TestPaths(unittest.TestCase):
    """Test suite for the ``oa::config::`` path helpers."""

    def test_data_dir(self):
        """Test that ``data_dir()` works as expected.

        It should resolve to the directory containing oa_config.
        """
        path = data_dir()
        self.assertEqual(_oa_config_dir, path)

    def test_library_dir(self):
        """Test that ``library_dir()`` works as expected.

        If should resolve to the directory containing oa_config.
        """
        path = library_dir()
        self.assertEqual(_oa_config_dir, path)


class TestVersionInfo(unittest.TestCase):
    """Test suite for ``oa::config::version_info``."""

    @staticmethod
    def config_verinfo() -> VersionInfo:
        """Return a ``VersionInfo`` from ``library_path()``."""
        return VersionInfo(library_path())

    def test_str(self):
        """Test that the ``VersionInfo`` can be created with a ``str``."""
        info = VersionInfo(str(library_path()))

    def test_no_default_ctor(self):
        """Test that the ``VersionInfo`` doesn't expose the default ctor."""
        with self.assertRaises(TypeError):
            _ = VersionInfo()

    def test_no_copy_ctor(self):
        """Test that the ``VersionInfo`` doesn't expose copy/move ctor."""
        with self.assertRaises(TypeError):
            _ = VersionInfo(self.config_verinfo())

    def test_company(self):
        """Test that the ``VersionInfo`` has the expected company string."""
        self.assertEqual("OA Developers", self.config_verinfo().company())

    def test_filename(self):
        """Test that the ``VersionInfo`` has the expected library file name."""
        self.assertEqual(library_path().name, self.config_verinfo().filename())

    def test_version_num(self):
        """Test that the ``VersionInfo`` version number is wrapped correctly."""
        self.assertTrue(isinstance(self.config_verinfo().version_num(), int))


if __name__ == "__main__":
    test_main(__name__)
